### ISSUE TITLE
Indicate artifact update problems in MR/PR title

### ISSUE THREAD

**Issue by languitar** [renovatebot/renovate#13483]
### What would you like Renovate to be able to do?

It would be nice if artifact update problems were indicated in the PR/MR titles. Despite update problems, pipeline might still succeed. Therefore, from the list of opened PRs you do not always have an overview of which PRs need manual care.

### If you have any ideas on how this should be implemented, please tell us here.

Adding a UTF-8 warning sign in case of artifact update issues might be sufficient.

### Is this a feature you are interested in implementing yourself?

No

**Comment by languitar** [comment]
We are on Gitlab. We see a comment on the MR, but that can easily be missed in the list of MRs and it also doesn't prevent merging

**Comment by rarkins** [comment]
Can you check that you have granted rights to create status checks to the Renovate user? Also can you capture the debug log for the branch when a commit with failing artifacts is done? e.g. you can take a branch which you know has failing artifacts and tick the retry/rebase button to force a new commit the next time it runs. There should be an attempt By Renovate to set a failing `renovate/artifacts` status check.

**Comment by languitar** [comment]
We are using a personal access token of a special user for renovate in our case. That  token has the API scope and the user is maintainer in every project. Yet, there's no status check. Also the debug log doesn't indicate anything inmportant:

```
DEBUG: Updating status check state to failed (repository=xxx, branch=renovate/xxx)
```

No error message afterwards that this failed.

Could it be that these checks don't work with detached merge request pipelines?

**Comment by rarkins** [comment]
Is there anything which Renovate could do differently when creating the status check which would mean that it would be visible?

**Comment by languitar** [comment]
There seems to be a concept of external status checks: https://docs.gitlab.com/ee/api/status_checks.html#get-project-external-status-checks But I am not sure if that applies for the periodic execution model of renovate because it seems external URL are calls for the checks.

I am not exactly sure, but maybe adding the [`ref` attribute](https://docs.gitlab.com/ee/api/commits.html#post-the-build-status-to-a-commit) to the POST request for the status could help: https://gitlab.com/gitlab-org/gitlab/-/issues/14064#possible-fixes

**Comment by languitar** [comment]
Interesting. If I manually add a status check to the affected commit now it becomes visible in the merge request pipeline. It seems one can also set `pipeline_id` to explicitly target the merge request pipeline.

**Comment by languitar** [comment]
Aaaah, the order of operations needs to changed. The MR must be created before setting the status check to fix this.

**Comment by rarkins** [comment]
In that case though you'd need prCreation=immediate (default)

**Comment by languitar** [comment]
We haven't touched that setting. Still, the debug message for setting the status appears before creating the MR.

**Comment by viceice** [comment]
It's probably a timing issue again, your gitlab is to slow to report pipeline status, so renovate posts first. I think we already have a delay, but that seems to low for you.

**Comment by languitar** [comment]
I have boiled down this problem to a close to minimal repository with a renovate configuration including an explicit `prCreation` with value `immediate`. Here are the relevant lines from the log file showing how PR creation and status check interleave:

```
 INFO: Branch created (repository=plan.one/experimental/renovate-cruft-tests/instance, branch=renovate/mocha-9.x)
       "commitSha": "8ba7871"
DEBUG: Got res with 0 results (repository=plan.one/experimental/renovate-cruft-tests/instance, branch=renovate/mocha-9.x)
DEBUG: Updating status check state to failed (repository=plan.one/experimental/renovate-cruft-tests/instance, branch=renovate/mocha-9.x)
DEBUG: Ensuring PR (repository=plan.one/experimental/renovate-cruft-tests/instance, branch=renovate/mocha-9.x)
DEBUG: There are 0 errors and 0 warnings (repository=plan.one/experimental/renovate-cruft-tests/instance, branch=renovate/mocha-9.x)
DEBUG: getBranchPr(renovate/mocha-9.x) (repository=plan.one/experimental/renovate-cruft-tests/instance, branch=renovate/mocha-9.x)
DEBUG: findPr(renovate/mocha-9.x, undefined, open) (repository=plan.one/experimental/renovate-cruft-tests/instance, branch=renovate/mocha-9.x)
DEBUG: Forcing PR because of artifact errors (repository=plan.one/experimental/renovate-cruft-tests/instance, branch=renovate/mocha-9.x)
DEBUG: Creating PR (repository=plan.one/experimental/renovate-cruft-tests/instance, branch=renovate/mocha-9.x)
       "prTitle": "Update dependency mocha to v9"
DEBUG: Creating Merge Request: Update dependency mocha to v9 (repository=plan.one/experimental/renovate-cruft-tests/instance, branch=renovate/mocha-9.x)
```

`Updating status check state to failed` appears before `Ensuring PR`.

**Comment by rarkins** [comment]
@languitar so is it your recommendation that Renovate should apply status checks only *after* creating the MR? Is there any documentation or issue in GitLab you've found which we can reference?

**Comment by languitar** [comment]
I am not aware of any explicit documentation on that. But with the merge request pipeline workflow the pipeline only gets created after the merge request has been created. So a status check for this pipeline can always only be set after the merge request was created.

**Comment by rarkins** [comment]
And gitlab completely ignores that there's a failing check for the *commit* which the MR points to? That seems like something which hopefully has a gitlab bug assigned or could be fixed

**Comment by languitar** [comment]
You can, in theory, configure completely different pipelines for individual commits/branches and the merge request based on that branch. Here's the relevant documentation: https://docs.gitlab.com/ee/ci/pipelines/merge_request_pipelines.html. In our case, we have completely disable pipelines for individual commits in favor of merge request pipelines and with that configuration I can somehow understand that the UI favors the merge request pipeline in merge requests while ignoring the commit pipeline which actually shouldn't exist at at all with our pipeline configuration.

It is a bit strange that the status check creates a commit pipeline that never existed before but from the Gitlab point of view I can also understand that there's no chance to associate the status check with a pipeline that doesn't exist at the point in time it is posted.

In any case: as this is probably the intended way Gitlab works, it would really be nice if there was a way to make renovate work with this model.

**Comment by rarkins** [comment]
OK, so perhaps we can alter our logic so that the status check setting is applied last, after the MR/PR has been created (assuming the PR is created immediately). Hopefully there's no problem with applying this logic to all platforms even though for now it seems like only an edge case in GitLab

**Comment by languitar** [comment]
> OK, so perhaps we can alter our logic so that the status check setting is applied last, after the MR/PR has been created 

That would actually have been my understanding of the "immediate" setting.

**Comment by rarkins** [comment]
No. Status checks as a concept apply to commits, not PRs. They are applied against SHAs. There would be no point delaying if it weren't for this gitlab limitation we are attempting to work around, and doing so will add undesirable complexity too

**Comment by languitar** [comment]
I would guess the Gitlab developers don't understand this as a limitation but as a feature ;-)

**Comment by dahaupt** [comment]
@languitar I just ran into this issue as well. Is there any workaround? I thought about using `prCreation="status-success"` temporarily in order to prevent auto merging while having open artifact update problems.

**Comment by languitar** [comment]
Sometimes it works now, sometimes it doesn't. I don't have a systematic workaround.

**Comment by dahaupt** [comment]
We are still facing this issue in our self-managed GitLab instance.

Is changing the order on `prCreation=immediate` a valid option? I'd be willing to create a PR for this.

I'm looking forward to an active discussion in order to at least get a workaround. Currently, we are dealing with invalid auto-merged package updates, which we have to revert manually.

**Comment by viceice** [comment]
you should configure the pipeline to fail on artifact update failures.

i think most tools support a flag to fail if there is a lockfile mismatched.

**Comment by viceice** [comment]
- https://github.com/renovatebot/renovate/pull/15971

that should probably also help

**Comment by chopeen** [comment]
> Aaaah, yes, the status check is correctly attached to the branch. But because of the merge request pipeline configuration, this check is completely invisible in the merge request.

I think I am failing a similar issue in GitLab:
- the main pipeline is an MR one and it succeeds :heavy_check_mark:,
- `renovate/artifacts` is created as a separate branch pipeline and this one fails :x:,
- all in all, MR is :heavy_check_mark: and GitLab does not prevent merging it.

Related discussion: https://github.com/renovatebot/renovate/discussions/19763#discussioncomment-4654216

**Comment by rarkins** [comment]
Could we have a feature like "transfer branch status to MR status" where if enabled, we check for any `renovate/*` status checks on the commit and then copy them to the MR?

Do you all use prCreation=immediate (default)?

**Comment by chopeen** [comment]
The proposed feature would solve my issue and yes, I use default value for the `prCreation` setting.
